questionpy_server/worker/worker/__init__.py

```diff
@@ -71,6 +74,10 @@ async def kill(self) -> None:
     def send(self, message: MessageToWorker) -> None:
         """Send a message to the worker."""
 
+    @abstractmethod
+    async def send_and_wait_for_response(self, message: MessageToWorker, expected_response_message: type[_M]) -> _M:
+        """Send a message and wait for a response of the given type."""
+
```

Comment:
`send_and_wait_for_response` habe ich primär für die Tests in die ABC hochgezogen und weil `send` auch public ist. Es wäre aber auch denkbar, beide privat zu machen und stattdessen eine Methode für `DebugExec` zu implementieren.

Reply:
Ich würde `send_and_wait_for_response` und `send` mit underscore als privat markieren, denn die sind ja eigentlich privat, soweit ich das überblicke.

Ich weiß nicht, ob ich `DebugExec` ganz verstehe, aber es existiert ja bloß für die Tests, um Code in den `WorkerManager` zu injizieren, oder?

Vielleicht wäre es besser den Production-Code (soweit möglich) frei von Testlogik zu halten. Dafür böte sich `monkeypatch` doch an. Dann könnte man auf `exec` und die `if __debug__`-Blöcke verzichten. Für den `SubprocessWorker` ist das slightly trickier, da man den natürlich nicht direkt monkeypatchen kann. Hab mal konkret skizziert, was ich meine: [api-changes-monkeypatch](https://github.com/questionpy-org/questionpy-server/commit/92b529530738b67cd0e89e10a8e83b6ddfb42a59)

Reply:
> Ich würde send_and_wait_for_response und send mit underscore als privat markieren, denn die sind ja eigentlich privat, soweit ich das überblicke.

Werden zumindest nicht von außen aufgerufen (bis auf den Test jetzt, aber das erledigt sich gleich), ja.

> Vielleicht wäre es besser den Production-Code (soweit möglich) frei von Testlogik zu halten. Dafür böte sich `monkeypatch` doch an. Dann könnte man auf `exec` und die `if __debug__`-Blöcke verzichten. Für den `SubprocessWorker` ist das slightly trickier, da man den natürlich nicht direkt monkeypatchen kann. Hab mal konkret skizziert, was ich meine: [api-changes-monkeypatch](https://github.com/questionpy-org/questionpy-server/commit/92b529530738b67cd0e89e10a8e83b6ddfb42a59)

Für den `ThreadWorker` hatte ich es ursprünglich ja auch so gemacht. (Nur eben `unittest.mock.patch`, nicht `pytest.monkeypatch`). Das hier war der Versuch, das TODO für den `SubprocessWorker` zu erledigen. Die Runtime-Main zu ersetzen ist eine super Idee, darauf war ich nicht gekommen. :+1: Werde ich in ähnlicher Form übernehmen. Ich würde allerdings bei `unittest.mock.patch` bleiben, zwecks Konsistenz. (Dass Produktivcode Teile enthält, die nur das Testen leichter machen, halte ich übrigens für normal und nicht "leaky".)

Reply:
In e835f6d59a85572ff43898b327c1b2aace06f69f. Schau dir das gern nochmal an. Ich habe mal ein `Co-authored-by` geschrieben, if you're fine with that.